Move MiqTask#human_status into it's own module

NickLaMuro opened the pull request:
Purpose or Intent
-----------------
To be used with Manageiq/manageiq#9559, this will allow the code for other models to reuse the `human_status` display logic, assuming they implement the methods `state` and `status`.  There is currently no enforcement requiring that those methods exist for this to be included in a class though.


Links
-----
* https://github.com/ManageIQ/manageiq/pull/9559


Steps for Testing/QA
--------------------
Added tests around the logic for `MiqTask::human_status`, so that should suffice for testing beyond checking in the app/UI in common functionality around MiqTasks still works.

NickLaMuro commented:
@miq-bot add_label refactoring

Fryguy commented:
Part of me thinks this is really a UI presenter concern.  Do you know where this is actually being used?  Perhaps this can just be a helper or in one of the presenter / decorators.

NickLaMuro commented:
@Fryguy Yup, I agree, but was unsure of proper way to go about refactoring it into more of a decorator without rocking the boat too much.

Basically, it seems to only be used in two places:

* `MiqReportResult#status`
* `MiqWidget#status`

So not very often, but the question is where those are now being used, and what they would affect if they were changed/removed.  Haven't looked into that, but I will report back with what I find.

kbrock commented:
Ugh, so the `{miq_report_result,miq_widget}#status` shows `task#human_status` which is made up from `task#status` and `task#state`

Is it possible for us to rename `{miq_report_result,miq_widget}#status` to `#human_status` so the field always means the same thing?

NickLaMuro commented:
@kbrock That is kind of what I am looking into right now.  Seeing if I can make `human_status` basically what is called by those two and changing out `status` and `state` of those classes to reference the aliased columns or delegate to the associated `miq_task`.  I need to see where `status` is being used first though before I make any more decisions, which is a bit of a exercise in project greping.

kbrock commented:
```bash
ag `[. ]status\b`
```

1411 matches across 368 `.rb` files

Wish I knew a better way to say something that is not in a string or comment.

GL

kbrock commented:
@NickLaMuro 

```ruby
def status
  # puts "called status", caller
  raise "called status"
end
```

```
bundle exec rake #or have travis do it
```

not fool proof, but may be a start

Fryguy commented:
So, taking a huge step back...what does this buy you...risk/reward/roi and all that?

kbrock commented:
@Fryguy we use "delegators" (with arel) for the 2 fields, bring them in as part of the main query and remove N+1 (and fetch less data)

not to mention `status` takes on a consistent meaning

NickLaMuro commented:
I just wanted to get this out in the open so we can make a decision on this (hopefully today) if it should included or not.

The patch at the bottom of this comment is from the combination of #9559 and a final modification to this PR, making human status work with configurable columns/methods (columns ideally).  Here are some quick numbers:


### Before

#### No Subtrees

Started GET "/report/explorer" for ::1 at 2016-07-12 17:03:49 -0500
Completed 200 OK in 4660ms (Views: 609.8ms | ActiveRecord: 594.0ms)

|      ms |       bytes |   objects | queries | query (ms) |   rows |
|    ---: |        ---: |      ---: |    ---: |       ---: |   ---: |
| 4,223.6 | N/A | 4,263,693 |     260 |    3,560.2 | 27,205 |


#### Single 10000+ node subtree loaded

Started GET "/report/explorer" for ::1 at 2016-07-12 17:11:49 -0500
Completed 200 OK in 45140ms (Views: 536.8ms | ActiveRecord: 6741.1ms)

|       ms |       bytes |    objects | queries | query (ms) |   rows |
|     ---: |        ---: |       ---: |    ---: |       ---: |   ---: |
| 44,799.6 | N/A | 63,164,707 |  20,824 |   16,718.7 | 47,769 |


### After

#### No Subtrees

Started GET "/report/explorer" for ::1 at 2016-07-12 16:48:33 -0500
Completed 200 OK in 4085ms (Views: 405.0ms | ActiveRecord: 594.9ms)

|      ms |       bytes |   objects | queries | query (ms) |   rows |
|    ---: |        ---: |      ---: |    ---: |       ---: |   ---: |
| 3,959.7 | N/A | 4,358,944 |     260 |    3,558.7 | 27,205 |


#### Single 10000+ node subtree loaded

Started GET "/report/explorer" for ::1 at 2016-07-12 16:55:46 -0500
Completed 200 OK in 18754ms (Views: 618.6ms | ActiveRecord: 730.6ms)

|       ms |      bytes |    objects | queries | query (ms) |   rows |
|     ---: |       ---: |       ---: |    ---: |       ---: |   ---: |
| 18,589.5 | N/A | 42,712,405 |     260 |    4,567.4 | 37,487 |


A few notes:

* The metrics from here are pre #9930, so the *All* of these metrics will be about 2secs faster, and 20,000ish rows cheaper.
* The are the slowest runs after a refresh on each one done in development mode, so some classes and queries should be eager loaded at this point (I think).
* The data can be similarly reproduced using [`miq_seeds`](https://github.com/NickLaMuro/miq_seeds) (see `README.md` for setup instructions), and the following seeding in the console:

```ruby
irb> require 'miq_seeds'
irb> seed_array = [10000, 7000, 6000, [400, 250, 60]*3, [20]*5, [10]*60 ].flatten
irb> MiqSeeds.generate :miq_reports, seed_array
```

Hopefully this helps answer @Fryguy 's question about risk/reward/ROI.  I tried spiking on doing this by moving the status stuff into the views/presenters, and found it to require a much bigger hammer (lots of things breaking) then just doing the following patch, which falls back to the old way of doing things when human status on `MiqReportResult` doesn't work (not using `::with_task_status_and_state`).


```
diff --git a/app/models/miq_report_result.rb b/app/models/miq_report_result.rb
index 96d6404..2a964f4 100644
--- a/app/models/miq_report_result.rb
+++ b/app/models/miq_report_result.rb
@@ -1,5 +1,6 @@
 class MiqReportResult < ApplicationRecord
   include_concern 'Purging'
+  include HumanStatus
 
   belongs_to :miq_report
   belongs_to :miq_group
@@ -16,6 +17,9 @@ class MiqReportResult < ApplicationRecord
 
   virtual_has_one :result_set,           :class_name => "Hash"
 
+  self.status_attribute = 'task_status'
+  self.state_attribute  = 'task_state'
+
   before_save do
     user_info = userid.to_s.split("|")
     if user_info.length == 1
@@ -31,7 +35,11 @@ class MiqReportResult < ApplicationRecord
   end
 
   def status
-    miq_task.nil? ? "Unknown" : miq_task.human_status
+    if miq_task_id.nil?
+      "Unknown"
+    else
+      (human_status rescue nil) || miq_task.human_status
+    end
   end
 
   def status_message
@@ -286,6 +294,22 @@ class MiqReportResult < ApplicationRecord
     result_type.to_sym == :html ? html_rows : report_results
   end
 
+  # Attempt to get the associated MiqTask status and state as an aliased column
+  # of task_status and task_state respectively
+  def self.with_task_status_and_state
+    tasks_table = MiqTask.arel_table
+
+    select(
+      arel_table[Arel.star],
+      tasks_table[:status].as('task_status'),
+      tasks_table[:state].as('task_state')
+    ).joins(
+      arel_table.join(tasks_table, Arel::Nodes::OuterJoin)
+                .on(arel_table[:miq_task_id].eq(tasks_table[:id]))
+                .join_sources
+    )
+  end
+
   def self.counts_by_userid
     where("userid NOT LIKE 'widget%'").select("userid, COUNT(id) as count").group("userid")
       .collect { |rr| {:userid => rr.userid, :count => rr.count.to_i} }
diff --git a/app/models/miq_task.rb b/app/models/miq_task.rb
index dedd8a1..f43fda2 100644
--- a/app/models/miq_task.rb
+++ b/app/models/miq_task.rb
@@ -1,15 +1,7 @@
 class MiqTask < ApplicationRecord
+  include HumanStatus
+
   serialize :context_data
-  STATE_INITIALIZED = 'Initialized'.freeze
-  STATE_QUEUED      = 'Queued'.freeze
-  STATE_ACTIVE      = 'Active'.freeze
-  STATE_FINISHED    = 'Finished'.freeze
-
-  STATUS_OK         = 'Ok'.freeze
-  STATUS_WARNING    = 'Warn'.freeze
-  STATUS_ERROR      = 'Error'.freeze
-  STATUS_TIMEOUT    = 'Timeout'.freeze
-  STATUS_EXPIRED    = 'Expired'.freeze
   validates_inclusion_of :state,  :in => [STATE_INITIALIZED, STATE_QUEUED, STATE_ACTIVE, STATE_FINISHED]
   validates_inclusion_of :status, :in => [STATUS_OK, STATUS_WARNING, STATUS_ERROR, STATUS_TIMEOUT]
 
@@ -138,23 +130,6 @@ class MiqTask < ApplicationRecord
     update_attributes(:state => STATE_FINISHED)
   end
 
-  def human_status
-    case state
-    when STATE_INITIALIZED then "Initialized"
-    when STATE_QUEUED      then "Queued"
-    when STATE_ACTIVE      then "Running"
-    when STATE_FINISHED
-      case status
-      when STATUS_OK      then "Complete"
-      when STATUS_WARNING then "Finished with Warnings"
-      when STATUS_ERROR   then "Error"
-      when STATUS_TIMEOUT then "Timed Out"
-      else raise _("Unknown status of: %{task_status}") % {:task_status => status.inspect}
-      end
-    else raise _("Unknown state of: %{task_status}") % {:task_status => state.inspect}
-    end
-  end
-
   def results_ready?
     status == STATUS_OK && !task_results.blank?
   end
diff --git a/app/presenters/tree_builder_report_saved_reports.rb b/app/presenters/tree_builder_report_saved_reports.rb
index c4e0271..769bd74 100644
--- a/app/presenters/tree_builder_report_saved_reports.rb
+++ b/app/presenters/tree_builder_report_saved_reports.rb
@@ -36,7 +36,9 @@ class TreeBuilderReportSavedReports < TreeBuilderReportReportsClass
   end
 
   def x_get_tree_custom_kids(object, count_only, _options)
-    scope = MiqReportResult.with_current_user_groups_and_report(from_cid(object[:id].split('-').last))
+    report_id = from_cid(object[:id].split('-').last)
+    scope = MiqReportResult.with_current_user_groups_and_report(report_id)
+                           .with_task_status_and_state
     count_only ? scope.size : scope.order("last_run_on DESC").to_a
   end
 end
```

Alternative solutions or ideas welcome.  Comment when available.

/cc @kbrock

NickLaMuro commented:
Basically covered by https://github.com/ManageIQ/manageiq/pull/11554, so closing
